src/TauBounds.cpp

```diff
@@ -46,14 +46,38 @@ void propagateLowerBound(int pop, double bound) {
 }
 
 
-void calculateTauUpperBounds() {
+void calculateTauUpperBounds1() {
   for (int gen = 0; gen < dataSetup.numLoci; gen++) {
     int rootNodeId = dataState.lociData[gen]->root;
     calculateLocusTauBounds(rootNodeId, gen);
   }
   propagateBoundsDownPopTree();
 }
 
+void calculateTauUpperBounds2() {
+  int mig, target, source;
+  double age;
+  for (int gen = 0; gen < dataSetup.numLoci; gen++) {
+    for (int i = 0; i < genetree_migs[gen].num_migs; i++) {
+      mig = genetree_migs[gen].living_mignodes[i];
+      target = genetree_migs[gen].mignodes[mig].target_pop;
+      source = genetree_migs[gen].mignodes[mig].source_pop;
+      age = genetree_migs[gen].mignodes[mig].age;
+      propagateUpperBoundDownwards(target, age);
+      propagateUpperBoundDownwards(source, age);
+    }
```

Comment:
same comment as with lower bounds. Here the bound is applied correctly to souce and target pops, but propagation should be done once and not every time you encounter a migration event. I also suggest that you propagate after ubound1 and ubound2 are computed. Maybe even have a single functiont hat propagates also the lower bounds?